File: packages/java/src/main/java/com/pezhwan/PezhwanClient.java
package com.pezhwan;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;

/**
 * Synchronous HTTP client for the PEZHWAN identity server.
 *
 * <p>Strings are decoded to the server's {@code {success, data, error}}
 * envelope; any non-2xx response raises {@link PezhwanApiException}.
 */
public final class PezhwanClient {

    private final String baseUrl;
    private final HttpClient http;
    private String accessToken;
    private String tenantId;

    private PezhwanClient(Builder builder) {
        this.baseUrl = stripTrailingSlash(builder.baseUrl);
        this.accessToken = builder.accessToken;
        this.tenantId = builder.tenantId;
        HttpClient.Builder cb = HttpClient.newBuilder()
                .connectTimeout(builder.timeout)
                .followRedirects(HttpClient.Redirect.NORMAL);
        if (builder.trustAllCerts) {
            cb.sslContext(InsecureTls.trustAllContext());
        }
        this.http = cb.build();

        this.auth = new AuthService(this);
        this.sessions = new SessionService(this);
        this.mfa = new MFAService(this);
        this.authorization = new AuthorizationService(this);
        this.tenant = new TenantService(this);
    }

    /** Builder for {@link PezhwanClient}. */
    public static final class Builder {
        private String baseUrl;
        private String accessToken;
        private String tenantId;
        private Duration timeout = Duration.ofSeconds(30);
        private boolean trustAllCerts = false;

        public Builder(String baseUrl) {
            this.baseUrl = baseUrl;
        }

        public Builder accessToken(String token) {
            this.accessToken = token;
            return this;
        }

        public Builder tenantId(String tenantId) {
            this.tenantId = tenantId;
            return this;
        }

        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }

        /** Disables TLS verification. For local/development hosts only. */
        public Builder trustAllCerts() {
            this.trustAllCerts = true;
            return this;
        }

        public PezhwanClient build() {
            return new PezhwanClient(this);
        }
    }

    public final AuthService auth;
    public final SessionService sessions;
    public final MFAService mfa;
    public final AuthorizationService authorization;
    public final TenantService tenant;

    /** Sets the bearer token used for subsequent requests. */
    public void setToken(String token) {
        this.accessToken = token;
    }

    public String getToken() {
        return accessToken;
    }

    /** Performs a GET and returns the decoded envelope data. */
    public Map<String, Object> get(String path) throws IOException, InterruptedException {
        return request("GET", path, null);
    }

    /** Performs a POST with a JSON body and returns the envelope data. */
    public Map<String, Object> post(String path, Map<String, Object> body)
            throws IOException, InterruptedException {
        return request("POST", path, body);
    }

    /** Performs a DELETE and returns the envelope data. */
    public Map<String, Object> delete(String path) throws IOException, InterruptedException {
        return request("DELETE", path, null);
    }

    private Map<String, Object> request(String method, String path, Map<String, Object> body)
            throws IOException, InterruptedException {
        HttpRequest.Builder builder = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .timeout(Duration.ofSeconds(45))
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .header("User-Agent", "pezhwan-java/0.1.0");
        if (accessToken != null && !accessToken.isEmpty()) {
            builder.header("Authorization", "Bearer " + accessToken);
        }
        if (tenantId != null && !tenantId.isEmpty()) {
            builder.header("X-Tenant-Id", tenantId);
        }

        String json = body == null ? null : JsonUtil.write(body);
        switch (method) {
            case "GET" -> builder.GET();
            case "DELETE" -> builder.DELETE();
            default -> builder.method(method, body == null
                    ? HttpRequest.BodyPublishers.noBody()
                    : HttpRequest.BodyPublishers.ofString(json));
        }

        HttpResponse<String> response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        Map<String, Object> payload = JsonUtil.read(response.body());
        int status = response.statusCode();
        if (status < 200 || status >= 300) {
            throw PezhwanException.PezhwanApiException.from(status, Models.Envelope.from(payload));
        }
        return Models.Envelope.from(payload).data();
    }

    /** Queries the public health endpoint. */
    public Map<String, Object> health() throws IOException, InterruptedException {
        return get("/v1/admin/health");
    }

    private static String stripTrailingSlash(String url) {
        return url == null || url.isEmpty() ? "" : url.replaceAll("/+$", "");
    }
}

File: packages/dotnet/Pezhwan/PezhwanClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;

namespace Pezhwan;

/// <summary>
/// Minimal HTTP client for the PEZHWAN identity server.
/// </summary>
public sealed class PezhwanClient
{
    private readonly HttpClient _http;
    private string _accessToken;
    private readonly string _tenantId;

    public PezhwanClient(string baseUrl, string? accessToken = null, string? tenantId = null,
        TimeSpan? timeout = null)
    {
        if (string.IsNullOrWhiteSpace(baseUrl))
            throw new ArgumentException("baseUrl is required", nameof(baseUrl));

        _http = new HttpClient
        {
            BaseAddress = new Uri(baseUrl.EndsWith("/") ? baseUrl : baseUrl + "/"),
            Timeout = timeout ?? TimeSpan.FromSeconds(30),
        };
        _http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        _http.DefaultRequestHeaders.UserAgent.ParseAdd("pezhwan-dotnet/0.1.0");
        _accessToken = accessToken ?? "";
        _tenantId = tenantId ?? "";

        Auth = new AuthService(this);
        Sessions = new SessionService(this);
        Mfa = new MFAService(this);
        Authorization = new AuthorizationService(this);
        Tenant = new TenantService(this);
    }

    public AuthService Auth { get; }
    public SessionService Sessions { get; }
    public MFAService Mfa { get; }
    public AuthorizationService Authorization { get; }
    public TenantService Tenant { get; }

    public string AccessToken
    {
        get => _accessToken;
        set => _accessToken = value ?? "";
    }

    /// <summary>Constructs a configured client with auth/session/mfa attached.</summary>
    public static PezhwanClient Create(string baseUrl, string? accessToken = null, string? tenantId = null,
        TimeSpan? timeout = null)
        => new PezhwanClient(baseUrl, accessToken, tenantId, timeout);

    /// <summary>Sends a JSON request returning the envelope's data payload.</summary>
    public async Task<Dictionary<string, object?>> RequestAsync(
        HttpMethod method, string path, object? body = null, CancellationToken ct = default)
    {
        using var request = NewRequest(method, path, body);
        using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, ct)
            .ConfigureAwait(false);

        var raw = await response.Content.ReadAsStringAsync(ct).ConfigureAwait(false);
        var payload = Parse(raw);
        if (!response.IsSuccessStatusCode)
            throw BuildException((int)response.StatusCode, payload);

        return payload.TryGetValue("data", out var data)
            && data is Dictionary<string, object?> map
                ? map
                : new Dictionary<string, object?>();
    }

    public Task<Dictionary<string, object?>> GetAsync(string path, CancellationToken ct = default)
        => RequestAsync(HttpMethod.Get, path, null, ct);

    public Task<Dictionary<string, object?>> PostAsync(string path, object? body = null, CancellationToken ct = default)
        => RequestAsync(HttpMethod.Post, path, body, ct);

    public Task<Dictionary<string, object?>> DeleteAsync(string path, CancellationToken ct = default)
        => RequestAsync(HttpMethod.Delete, path, null, ct);

    /// <summary>Queries the public health endpoint.</summary>
    public Task<Dictionary<string, object?>> HealthAsync(CancellationToken ct = default)
        => GetAsync("/v1/admin/health", ct);

    /// <summary>Fetches the OIDC discovery document.</summary>
    public Task<Dictionary<string, object?>> DiscoveryAsync(CancellationToken ct = default)
        => GetAsync("/.well-known/openid-configuration", ct);

    private HttpRequestMessage NewRequest(HttpMethod method, string path, object? body)
    {
        var request = new HttpRequestMessage(method, path.TrimStart('/'));
        if (!string.IsNullOrEmpty(_accessToken))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _accessToken);
        if (!string.IsNullOrEmpty(_tenantId))
            request.Headers.TryAddWithoutValidation("X-Tenant-Id", _tenantId);
        if (body is not null)
            request.Content = JsonContent.Create(body);
        return request;
    }

    private static Dictionary<string, object?> Parse(string raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
            return new Dictionary<string, object?>();
        try
        {
            return JsonSerializer.Deserialize<Dictionary<string, object?>>(raw)
                ?? new Dictionary<string, object?>();
        }
        catch (JsonException)
        {
            return new Dictionary<string, object?>();
        }
    }

    private static PezhwanApiException BuildException(int status, Dictionary<string, object?> payload)
    {
        var code = payload.TryGetValue("error", out var e)
                   && e is Dictionary<string, object?> error
                   && error.TryGetValue("code", out var c)
            ? c?.ToString() ?? "UNKNOWN"
            : "UNKNOWN";
        var message = payload.TryGetValue("error", out var m)
                      && m is Dictionary<string, object?> err2
                      && err2.TryGetValue("message", out var msg)
            ? msg?.ToString() ?? ""
            : ((HttpStatusCode)status).ToString();
        return new PezhwanApiException(status, code, message);
    }
}